SingleHost + TLS on OpenShift

sleshchenko opened the pull request:
### What does this PR do?
This PR enables Single host + TLS on OpenShift for basic routing.
It has self-describing commits.

:warning: It worths noting that changes are done on creator access only.
Now creator access only access is provided only for workspaces annotated with `controller.devfile.io/restricted-access: true`
Now CloudShell workspace (not web-terminal) can be opened only with OpenShift OAuth routing.
CloudShell workspace also does not support single-host feature because it uses absolute references at this point.

### TODOs:

- [ ] Theia /che/cluster-ip is failing because it uses absolute reference. Artem will adapt that code to single-host;
**Theia Webview**
- [x] Theia Webview does not work on Chrome if you don't import CA into the browser. Artem will create an issue to expose this info and the most probable solution is just documenting it, or improve Theia to render the proper error page;
- [x] Theia Webview is not stable
![Screenshot_20201201_145712](https://user-images.githubusercontent.com/5887312/100743924-a729e480-33e5-11eb-853c-c5303ddcafbf.png)

```
2020-12-01 14:53:12.187 root ERROR Security problem: Unable to configure separate webviews domain:  Params: Error: Unable to get workspace containers. Cause: [object Object]
    at CheServerWorkspaceServiceImpl.<anonymous> (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:193:31)
    at step (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:62:23)
    at Object.throw (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:43:53)
    at rejected (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:35:65)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```
It works depending on the container startup sequence.
Theia needs to try again when Che is not available for such fundamental stuff like Webview or che-api-sidecar may be moved to a dedicated pod which we start first, but no one of them worths attention because Che API Sidecar is just a temporary solution and it should be reworked in a proper way where Theia access like K8s cluster to get needed info;

- [x] OpenShift OAuth expects to be the only application on the host and sends an absolute redirect
![Screenshot_20201201_145951](https://user-images.githubusercontent.com/5887312/100744098-f112ca80-33e5-11eb-9240-eadfeb58750f.png)

- [ ] Endpoint name is not unique. Should we include the component name to path rule as well?

### What issues does this PR fix or reference?
https://github.com/devfile/devworkspace-operator/issues/200

### Is it tested? How?

#### Minikube
Verified that on minikube everything works as previously + now terminal works fine for:
```bash
make install

kubectl apply -f ./samples/theia-next.yaml

kubectl apply -f ./samples/all-in-one-theia-nodejs.devworkspace.yaml
```

#### CRC
Verified that on crc now:
- basic routing uses single-host approach;
- Theia works as expected (apart from Webview which is not stable and it's known issue described above);
- now terminal works fine;
for:
```bash
kubectl apply -f ./samples/theia-next.yaml

kc apply -f ./samples/all-in-one-theia-nodejs.devworkspace.yaml
```
![Screenshot_20201127_173303](https://user-images.githubusercontent.com/5887312/100464789-e222d400-30d6-11eb-842a-36cda948bc3e.png)

CloudShell works as it used to work with host per endpoint approach.

sleshchenko commented:
/test v5-devworkspaces-operator-e2e

amisevsk reviewed (approved):
A few minor comments below, but LGTM at its core. I'll test it later today, once I get `crc` up and running

amisevsk commented on controllers/controller/workspacerouting/solvers/resolve_endpoints.go:
```diff
@@ -75,16 +75,15 @@ func resolveURLForEndpoint(
 		Scheme: string(protocol),
 		Host:   host,
-		Path:   path,
+		Path:   path + endpoint.Path,
```
Also, does it make sense to ensure path is formatted correctly here? E.g. we could use `path.Join()` here

sleshchenko replied:
> Also, does it make sense to ensure path is formatted correctly here? E.g. we could use path.Join()

Sure, it's in TODO list )

sleshchenko commented on controllers/controller/workspacerouting/solvers/openshift_oauth_solver.go:
```diff
@@ -102,25 +102,12 @@ func (s *OpenShiftOAuthSolver) getProxyRoutes(
 			endpoint := proxyEndpoint.publicEndpoint
-			var tls *routeV1.TLSConfig = nil
-			if endpoint.Secure {
-				if endpoint.Attributes.GetString(string(controllerv1alpha1.TYPE_ENDPOINT_ATTRIBUTE), nil) == "terminal" {
```
I don't. I assume that code worked someday when terminal was not excluded from secure servers. But now this code can't be reached and I believe there is no reason to handle terminal differently

sleshchenko commented on samples/theia-next.yaml:
```diff
@@ -0,0 +1,26 @@
+  name: theia
+spec:
+  started: true
+  routingClass: 'basic'
```
I'll remove it from Theia devfiles I've added.
But will keep for CloudShell

sleshchenko commented on samples/cloud-shell.yaml:
```diff
@@ -2,8 +2,13 @@ kind: DevWorkspace
 metadata:
   name: cloud-shell
+  annotations:
+    # it's important to make workspace immutable to make sure nobody with right access
```
Since we add this annotation in four places, I think it makes sense to remove these comments from our samples.

JPinkney replied:
Yeah, I like the idea of putting these annotations in the README

sleshchenko replied:
I've added the corresponding section into README.md. I've used the format as https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/ has. Feedback or remarks are welcome.

sleshchenko commented on controllers/controller/workspacerouting/solvers/openshift_oauth_solver.go:
```diff
@@ -102,25 +102,12 @@ func (s *OpenShiftOAuthSolver) getProxyRoutes(
-			}
 			route := getRouteForEndpoint(endpoint, workspaceMeta)
-			route.Spec.TLS = tls
-			if route.Annotations == nil {
-				route.Annotations = map[string]string{}
```
In general, it makes code safer. In fact, it never is nil since this route is got from func that always assign annotations.
I'll revert this check

sleshchenko commented on controllers/controller/workspacerouting/solvers/common.go:
```diff
@@ -95,22 +94,25 @@ func getServicesForEndpoints(endpoints map[string]controllerv1alpha1.EndpointLis
-			Labels: map[string]string{
-				config.WorkspaceIDLabel: meta.WorkspaceId,
+	var services []corev1.Service
+	services = append(services)
```
Tiny clean up that itself needs cleaning up :-D

JPinkney commented on controllers/controller/workspacerouting/workspacerouting_controller.go:
```diff
@@ -79,7 +79,7 @@ func (r *WorkspaceRoutingReconciler) Reconcile(req ctrl.Request) (ctrl.Result, e
 
 	if instance.Status.Phase == controllerv1alpha1.RoutingFailed {
-		return reconcile.Result{}, err
+		return reconcile.Result{}, nil
```
Not sure if I missed something, but why are we no longer returning an error here?

sleshchenko replied:
I've removed err from here because it's confusing.
This err variable is declared in another place, and if any error occurred - it's thrown much earlier.
Here it's always nil. So, I'm returning nil here to make it clearer that when routing is failing we do nothing.

amisevsk commented on pkg/common/naming.go:
```diff
@@ -43,6 +43,20 @@ func EndpointHostname(workspaceId, endpointName string, endpointPort int, routin
+}
+
+func EndpointPath(endpointName string) string {
+	return "/" + endpointName + "/"
```
In testing on `crc` I found a few issues with our approach to paths:
- When we suffix the path with `/`, that exact string must be used to access the endpoint. This is an issue because the ideURL ends up without the trailing `/` so clicking the ideURL takes you to the OpenShift "application not available" page.
- The workspace never enters a `Running` state -- it's stuck in `Starting`. I believe (but haven't verified) that this is because we use `ideURL` in `checkServerStatus()`: 
    ```go
    ideUrl := workspace.Status.IdeUrl
    healthz, err := url.Parse(ideUrl)
    healthz.Path = healthz.Path + "healthz"
    ```
    since I'm seeing `IdeURL` as e.g. https://workspace99cfef48974c4fef.apps-crc.testing/theia it would appear that we're trying to ping `https://workspace99cfef48974c4fef.apps-crc.testing/theiahealthz`, which returns 503 instead of a 4xx that we're expecting. This causes the readiness check to fail at the end.

sleshchenko replied:
ideUrl was with trailing `/`... probably it went after Path.join stuff
Interesting that theia is not loaded when you access it without a trailing slash, seems Theia needs the right redirect:
without trailing slash
![Screenshot_20201201_142351](https://user-images.githubusercontent.com/5887312/100740201-fcafc280-33e0-11eb-974d-c04fa2603f37.png)
with trailing slash
![Screenshot_20201201_142340](https://user-images.githubusercontent.com/5887312/100740205-fe798600-33e0-11eb-8631-3b1c406fbb78.png)

So, I'll make sure that endpoints have trailing slash if it's path rule, and it should be possible to specify endpoint path without a slash.

sleshchenko replied:
Should be fixed with https://github.com/devfile/devworkspace-operator/pull/201/commits/db048ad7543389f7c9d57e4b279ffd418b1e5e6f

amisevsk reviewed (commented):
I ran into a few issues on `crc`. I'm not sure if this is expected at this point, but during theia startup  I see the error message 
> 2020-11-30 22:03:44.880 root ERROR Security problem: Unable to configure separate webviews domain:  Params: Error: Unable to get workspace containers. Cause: [object Object]


logged. This may also be related to the issue below.

sleshchenko commented:
/test v5-devworkspaces-operator-e2e

sleshchenko commented:
Unreal, even e2e tests passed :-D

amisevsk reviewed (commented):
Testing on `crc` + firefox and chrome, I can start up `theia-next` and see the webview (no images are loaded). However, once Theia is loaded, I see a tight loop on requests to `https://workspace918799a646044762.apps-crc.testing/che/client-ip` which all return 503 -- this route does not exist on the cluster.

I don't know enough about theia internals to debug further, and the theia logs become enormous pretty quickly with an (probably related) error 
```
Promise rejection not handled in one second: Error: Request failed with status code 503 , reason: Error: Request failed with status code 503
With stack trace: UJUX/e.exports@https://workspace918799a646044762.apps-crc.testing/theia/vendors.c8bdf28598a14f9b5776.js:212:2276
MkyZ/e.exports@https://workspace918799a646044762.apps-crc.testing/theia/vendors.c8bdf28598a14f9b5776.js:116:5450
hR+n/e.exports/</p.onreadystatechange@https://workspace918799a646044762.apps-crc.testing/theia/vendors.c8bdf28598a14f9b5776.js:293:2425 
2020-12-01 18:56:12.186 root ERROR [hosted-plugin: 47] Promise rejection not handled in one second: Error: Request failed with status code 503 , reason: Error: Request failed with status code 503
With stack trace: UJUX/e.exports@https://workspace918799a646044762.apps-crc.testing/theia/vendors.c8bdf28598a14f9b5776.js:212:2276
MkyZ/e.exports@https://workspace918799a646044762.apps-crc.testing/theia/vendors.c8bdf28598a14f9b5776.js:116:5450
hR+n/e.exports/</p.onreadystatechange@https://workspace918799a646044762.apps-crc.testing/theia/vendors.c8bdf28598a14f9b5776.js:293:2425
```

amisevsk commented:
For 
> It works depending on the container startup sequence.
> Theia needs to try again when Che is not available for such fundamental stuff like Webview or che-api-sidecar may be moved to a dedicated pod which we start first, but no one of them worths attention because Che API Sidecar is just a temporary solution and it should be reworked in a proper way where Theia access like K8s cluster to get needed info;

it may just be a bit of weird trivia, but I think we can get around this by making sure che-rest-apis is the first container in the pod spec. Kube starts containers in the order they're listed, iirc.

sleshchenko commented:
> I don't know enough about theia internals to debug further, and the theia logs become enormous pretty quickly with an (probably related) error

Known bug, Artem will take a look at that. See PR description )

amisevsk commented:
> Known bug, Artem will take a look at that. See PR description )

Ah missed that line, apologies.

sleshchenko commented:
New happy path check helps me to figure out that creator-access only is broken. I'll take a look tomorrow on that

sleshchenko commented:
Creator access only is recovered in https://github.com/devfile/devworkspace-operator/pull/201/commits/d5368c8706a74b754df687d60532fec33c13fa10
I just missed restrictedAnnotation propagation to deployment and pod.

Also, I think it may be redundant to deny workspace update for workspace with not-restricted-access, so it's done in https://github.com/devfile/devworkspace-operator/pull/201/commits/5dc1a9f79b843f414d9f116b20200a3d08650885

Note that all workspace-related objects, route, service, deployment are still immutable for both workspace types (with and without restricted access). We may keep it only for restricted access related workspace if we decide, but it may be done in a separate PR as well.

sleshchenko commented:
/test v5-devworkspaces-operator-e2e

sleshchenko commented:
![Screenshot_20201202_143033](https://user-images.githubusercontent.com/5887312/100872744-f423be80-34aa-11eb-99c2-be2b138a7920.png)

sleshchenko commented:
/test v5-devworkspaces-operator-e2e

JPinkney commented:
Just tested on cluster bot cluster and ran into a few things:

##### For samples/theia-next.yaml
Cloning when using `samples/theia-next.yaml` failed with `Couldn't clone https://github.com/che-samples/web-nodejs-sample.git: fatal: could not create work tree dir '/projects/project': Permission denied`. Other than the project not being available I was able to get into the theia terminal and the workspace view was working

##### For samples/all-in-one-theia-nodejs.devworkspace.yaml
Seems like i'm getting errors when its trying to find other containers so the workspace view isn't working :

```
2020-12-02 20:17:02.405 root ERROR Error: Unable to get workspace containers. Cause: [object Object]
    at CheServerWorkspaceServiceImpl.<anonymous> (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:184:31)
    at step (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:53:23)
    at Object.throw (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:34:53)
    at rejected (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:26:65)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
2020-12-02 20:17:02.491 root ERROR Unable to create Che API REST Client: "CHE_WORKSPACE_TELEMETRY_BACKEND_PORT" is not set.
2020-12-02 20:17:02.526 root ERROR Request getCurrentWorkspacesContainers failed with error: Unable to get workspace containers. Cause: [object Object] Params: Error: Unable to get workspace containers. Cause: [object Object]
    at CheServerWorkspaceServiceImpl.<anonymous> (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:184:31)
    at step (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:53:23)
    at Object.throw (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:34:53)
    at rejected (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:26:65)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
2020-12-02 20:17:02.529 root ERROR Request findUniqueEndpointByAttribute failed with error: Unable to get workspace containers. Cause: [object Object] Params: Error: Unable to get workspace containers. Cause: [object Object]
    at CheServerWorkspaceServiceImpl.<anonymous> (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:184:31)
    at step (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:53:23)
    at Object.throw (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:34:53)
    at rejected (/home/theia/node_modules/@eclipse-che/theia-remote-api/lib/node/che-server-workspace-service-impl.js:26:65)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

![2020-12-02-152220_1500x968_scrot](https://user-images.githubusercontent.com/8839537/100927295-b2971180-34b2-11eb-8a57-93700dcb7cea.png)

The workspace view isn't populated, project didn't clone etc. Not exactly sure why though

sleshchenko commented:
After I found and fixed bug with empty annotations (I had introduced before), AllInOne Theia and Theia Next work just fine on OpenShift 4.7 powered by cluster bot.
![Screenshot_20201203_111915](https://user-images.githubusercontent.com/5887312/100990689-d877f100-355a-11eb-9f32-278f8008bef4.png)
and cloud shell as well.

sleshchenko commented:
I assume permissions issues could be related to mkdir which currently implemented in a wrong way, and it's actual for master branch as well.
Theia is not able to access to Che API Sidecar, that's strange and should be investigated more if we can reproduce it, but I don't see how it may be related to that PR.

sleshchenko commented:
There are no critical objections against this PR, so, I'm merging it not to inflate this PR even more. I will be happy to address any feedback with separate PRs if any.

sleshchenko commented:
:man_facepalming: I forgot to push made fixes, so they are merged in https://github.com/devfile/devworkspace-operator/pull/213

sleshchenko commented:
> it may just be a bit of weird trivia, but I think we can get around this by making sure che-rest-apis is the first container in the pod spec. Kube starts containers in the order they're listed, iirc. If che-rest-apis is temporary this may be a worthwhile workaround.

Quickly tried and it does not help 
![Screenshot_20201208_112835](https://user-images.githubusercontent.com/5887312/101465378-8a8f2e80-3948-11eb-9378-b4d2a1b92ebe.png)

sleshchenko commented:
I was quick in my conclusion. The error above is caused by Theia 7.20 which has issues with single host. Seems it helps to workaround an issue and I'll create PR
